lib/spack/spack/config.py

```diff
@@ -207,61 +208,349 @@ def __repr__(self):
+                data = data[key]
+            data[parts[0]] = value
+
+            self.update_config(section, section_data, scope=scope)
+
+    def __iter__(self):
+        """Iterate over scopes in this configuration."""
+        for scope in self.scopes.values():
+            yield scope
+
+    def print_section(self, section):
+        """Print a configuration to stdout."""
+        try:
+            data = syaml.syaml_dict()
+            data[section] = self.get_config(section)
+            syaml.dump(data, stream=sys.stdout, default_flow_style=False)
+        except (yaml.YAMLError, IOError):
+            raise ConfigError("Error reading configuration: %s" % section)
+
+
+@contextmanager
+def override(path, value):
+    """Simple way to override config settings within a context."""
+    overrides = InternalConfigScope('overrides')
+
+    cfg = config()
+    cfg.push_scope(overrides)
+    cfg.set(path, value, scope='overrides')
+
+    yield cfg
+
+    scope = cfg.pop_scope()
+    assert scope is overrides
+
+
+def config():
+    """Singleton Configuration instance.
+
+    This constructs one instance associated with this module and returns
+    it. It is bundled inside a function so that configuratoin can be
+    initialized lazily.
+
+    Return:
+        (Configuration): object for accessing spack configuration
+
+    """
+    global _configuration
+    if not _configuration:
+        _configuration = Configuration()
+
+        # first do the builtin, hardcoded defaults
+        defaults = InternalConfigScope('_builtin', config_defaults)
+        _configuration.push_scope(defaults)
+
+        # Each scope can have per-platfom overrides in subdirectories
+        platform = spack.architecture.platform().name
+
+        # add each scope and its platform-specific directory
+        for name, path in configuration_paths:
+            _configuration.push_scope(ConfigScope(name, path))
+
+            plat_name = '%s/%s' % (name, platform)
+            plat_path = os.path.join(path, platform)
+            _configuration.push_scope(ConfigScope(plat_name, plat_path))
+
+        # we make a special scope for spack commands so that they can
+        # override configuration options.
+        _configuration.push_scope(InternalConfigScope('command_line'))
+
+    return _configuration
+
+
+#: This is the global singleton configuration for Spack.
+#: TODO: consider making this NOT global and associate it with a spack instance
+_configuration = None
+
+
+def get(path, default=None, scope=None):
+    """Module-level wrapper for ``Configuration.get()``."""
+    return config().get(path, default, scope)
+
+
+def set(path, value, scope=None):
+    """Convenience function for getting single values in config files.
+
+    Accepts the path syntax described in ``get()``.
+    """
+    return config().set(path, value, scope)
+
+
+def scopes():
+    """Convenience function to get list of configuration scopes."""
+    return config().scopes
```

Comment:
Why do we need wrappers to mask the singleton? Can't we avoid this?

Reply:
It's not easy to put property semantics at module scope.  You can, but IMHO it's not worth the effort.

Reply:
My preference is consistency. Earlier we had

`config.get('packages:...')`

and 

`config.get_configuration().scopes()`

(`get_configuration()` has since been renamed to `config()`). Having two different ways of interacting with the singleton was confusing IMO. I think masking the singleton is OK as long as it's clear what is going on when someone looks at `config.py` specifically.